azurelinuxagent/common/protocol/wire.py

```diff
@@ -623,103 +596,141 @@ def call_storage_service(http_req, *args, **kwargs):
-        for version_uri in version_uris_shuffled:
-
-            if datetime.now() - start_time > timedelta(minutes=timeout_in_minutes, milliseconds=timeout_in_ms):
-                logger.warn("Agent timed-out after {0} minutes while fetching extension manifests. {1}/{2} uris tried.",
-                    timeout_in_minutes, uris_tried, len(version_uris))
-                break
-
-            # GA expects a location and failoverLocation in ExtensionsConfig, but
-            # this is not always the case. See #1147.
-            if version_uri is None:
-                logger.verbose('The specified manifest URL is empty, ignored.')
-                continue
-
-            # Disable W0640: OK to use version_uri in a lambda within the loop's body
-            direct_func = lambda: self.fetch(version_uri)[0]  # pylint: disable=W0640
-            # NOTE: the host_func may be called after refreshing the goal state, be careful about any goal state data
-            # in the lambda.
-            # Disable W0640: OK to use version_uri in a lambda within the loop's body
-            host_func = lambda: self.fetch_manifest_through_host(version_uri)  # pylint: disable=W0640
+
+        uris_shuffled = uris
+        random.shuffle(uris_shuffled)
+        most_recent_error = "None"
+
+        for index, uri in enumerate(uris_shuffled):
+            elapsed = datetime.now() - start_time
+            if elapsed > _DOWNLOAD_TIMEOUT:
+                message = "Timeout downloading {0}. Elapsed: {1} URIs tried: {2}/{3}. Last error: {4}".format(download_type, elapsed, index, len(uris), ustr(most_recent_error))
+                raise ExtensionDownloadError(message, code=ExtensionErrorCodes.PluginManifestDownloadError)
 
             try:
-                manifest = self.send_request_using_appropriate_channel(direct_func, host_func)
-                if manifest is not None:
-                    host = self.get_host_plugin()
-                    host.update_manifest_uri(version_uri)
-                    return manifest
-            except Exception as error:
-                logger.warn("Failed to fetch manifest from {0}. Error: {1}", version_uri, ustr(error))
+                # Disable W0640: OK to use uri in a lambda within the loop's body
+                response = self._download_using_appropriate_channel(lambda: direct_download(uri), lambda: hgap_download(uri))  # pylint: disable=W0640
 
-            uris_tried += 1
+                if on_downloaded():
+                    return uri, response
 
-        raise ExtensionDownloadError("Failed to fetch manifest from all sources")
+            except Exception as exception:
+                most_recent_error = exception
 
-    def stream(self, uri, destination, headers=None, use_proxy=None, max_retry=None):
+        raise ExtensionDownloadError("Failed to download {0} from all URIs. Last error: {1}".format(download_type, ustr(most_recent_error)), code=ExtensionErrorCodes.PluginManifestDownloadError)
```

Comment:
I extended the error message to also include the errors from the direct and HGAP channels